### PR TITLE
Add webhook support

### PR DESCRIPTION
**Pull request by StrongMonkey**

### CODE REVIEW

**Comment by StrongMonkey** [comment]
https://github.com/rancher/fleet/issues/252

**Comment by StrongMonkey** [comment]
@nickgerace Can you help review this PR? For adding webhook support in fleet

File: pkg/webhook/webhook.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+func New(ctx context.Context, rContext *types.Context) *Webhook {
+	webhook := &Webhook{
+		gitjobs:   rContext.Gitjob.Gitjob().V1().GitJob(),
+		secrets:   rContext.Core.Core().V1().Secret(),
+		namespace: rContext.Namespace,
+	}
+
+	rContext.Core.Core().V1().Secret().OnChange(ctx, "webhook-secret", webhook.onSecretChange)
```
Comment (nickgerace) [inline pkg/webhook/webhook.go:54]: Does it matter that `webhook-secret` is hardcoded here?

File: pkg/webhook/webhook.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+func HandleHooks(ctx context.Context, rContext *types.Context) http.Handler {
+	root := mux.NewRouter()
+	webhook := New(ctx, rContext)
+	root.UseEncodedPath()
+	root.Handle("/", webhook)
+	aggregation.Watch(ctx,
+		rContext.Core.Core().V1().Secret(),
+		rContext.Namespace,
+		"steve-aggregation",
```
Comment (nickgerace) [inline pkg/webhook/webhook.go:216]: Same question here about hardcoing

File: pkg/webhook/webhook.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	}
+	rw.WriteHeader(200)
+	rw.Write([]byte("succeeded"))
+}
+
+func logAndReturn(rw http.ResponseWriter, err error) {
+	logrus.Errorf("Webhook processing failed: %s", err)
+	rw.Write([]byte(err.Error()))
+	rw.WriteHeader(500)
+	return
```
Comment (nickgerace) [inline pkg/webhook/webhook.go:205]: Why is the return necessary?

File: pkg/webhook/webhook.go [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+	if secret == nil || secret.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if secret.Name != webhookSecretName && secret.Namespace != w.namespace {
+		return nil, nil
+	}
+
+	var err error
+	w.github, err = github.New(github.Options.Secret(string(secret.Data[githubKey])))
```
Comment (nickgerace) [inline pkg/webhook/webhook.go:68]: What happens when a user does not have one of these keys?

File: pkg/webhook/webhook.go (hunk 2 above)
Comment (StrongMonkey) [reply pkg/webhook/webhook.go:216]: Same, handler name.

File: pkg/webhook/webhook.go (hunk 4 above)
Comment (StrongMonkey) [reply pkg/webhook/webhook.go:68]: Good question. I checked all the function and it doesn't return error.
```
func (WebhookOptions) Secret(secret string) Option {
	return func(hook *Webhook) error {
		hook.secret = secret
		return nil
	}
}
```

If user doesn't have the key, we just don't validate the payload. This means you have no protection against payload from world. Also see https://argoproj.github.io/argo-cd/operator-manual/webhook/

Comment (nickgerace) [reply pkg/webhook/webhook.go:68]: Ah I see. I like the implementation. It was not obvious at first, but that makes sense.

**Review by nickgerace — APPROVED** [review]
LGTM after nits

**Pull request merged**